include/envoy/router/router.h

```diff
@@ -957,6 +957,44 @@ class Route {
 
 using RouteConstSharedPtr = std::shared_ptr<const Route>;
 
+/**
+ * RouteCallback, returns one of these enums to the route matcher to indicate
+ * if the matched route has been accepted or it wants the route matching to
+ * continue.
+ */
+enum class RouteMatchStatus {
+  // Continue matching route
+  Continue,
+  // Accept matched route
+  Accept
+};
+
+/**
+ * RouteCallback, is passed this enum to indicate if more routes are available for evaluation.
+ */
+enum class RouteEvalStatus {
+  // Has more routes that can be evaluated for match
+  HasMoreRoutes,
+  // All routes have been evaluated for match
+  NoMoreRoutes
+};
+
+/**
+ * RouteCallback can be used to override routing decision made by the Route::Config::route,
+ * this callback is passed the RouteConstSharedPtr, when a matching route is found, and
+ * RouteEvalStatus indicating whether there are more routes available for evaluation.
+ *
+ * RouteCallback will be called back only when at least one matching route is found, if no matching
+ * routes are found RouteCallback will not be invoked. RouteCallback can return one of the
+ * RouteMatchStatus enum to indicate if the match has been accepted or should the route match
+ * evaluation continue.
+ *
+ * Returning RouteMatchStatus::Continue, when no more routes available for evaluation will result in
+ * no further callbacks and no route is deemed to be accepted and nullptr is returned to the caller
+ * of Route::Config::route.
+ */
+using RouteCallback = std::function<RouteMatchStatus(RouteConstSharedPtr, RouteEvalStatus)>;
```

Comment:
could you clarify what's the expected behavior here if the callback throws an exception?

Reply:
Behavior would be same as any filter throwing exception